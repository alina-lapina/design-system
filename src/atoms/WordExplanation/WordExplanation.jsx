import React, { useEffect, useState, useRef } from 'react';
import PropTypes from 'prop-types';
import { FileText } from 'react-feather';

const WordExplanation = ({
	explanation, children,
}) => {
	const node = useRef();
	const infoContainer = useRef();
	const [open, setOpen] = useState(false);
	const [position, setPosition] = useState({ x: 0, y: 0 });

	const handleClickOutside = e => {
		if (node.current.contains(e.target)) {
			return;
		}
		setOpen(false);
	};

	const calculatePosition = (x, y) => {
		const topMargin = 30;
		const infoBoxWidth = 300;
		if (open && window.innerWidth <= 768) {
			// Center for mobile devices
			setPosition({ x: (window.innerWidth - infoContainer.current.clientWidth) / 2, y: y + topMargin });
		}
		if (open && window.innerWidth > 768) {
			if (x + infoBoxWidth > window.innerWidth) {
				// Detect if box will go outside of screen
				const adjustment = x + infoBoxWidth - window.innerWidth;
				setPosition({ x: x - adjustment, y: y + topMargin });
			} else {
				setPosition({ x, y: y + topMargin });
			}
		}
	};

	useEffect(() => {
		calculatePosition(node.current.offsetLeft, node.current.offsetTop);
		if (open) {
			document.addEventListener('mousedown', handleClickOutside);
		} else {
			document.removeEventListener('mousedown', handleClickOutside);
		}

		return () => {
			document.removeEventListener('mousedown', handleClickOutside);
		};
	}, [open]);

	return (
		<span ref={node} onClick={() => setOpen(!open)} className="word-explanation-wrap">
			{children}<FileText size="12" />
			<div className="animate-background" />
			{open && (
				<div className="info-box-wrapper" ref={infoContainer} style={{ top: position.y, left: position.x }}>
					<div className="arrow">
						<svg width="16" height="16">
							<rect width="16" height="16" rotate="45deg" />
						</svg>
					</div>
					<div className="info-box">
						<i className="close-button">x</i>
						<span className="info-text">{explanation}</span>
					</div>
				</div>
			)}
		</span>
	);
};

WordExplanation.defaultProps = {};

WordExplanation.propTypes = {
	children: PropTypes.oneOfType([
		PropTypes.element,
		PropTypes.string,
	]),
	explanation: PropTypes.string.isRequired,
};

export default WordExplanation;
