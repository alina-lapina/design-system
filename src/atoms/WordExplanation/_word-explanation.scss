.word-explanation-wrap {
  cursor: pointer;
  position: relative;
  text-decoration: dotted underline;
  text-decoration-color: $ssb-green-4;

  .animate-background {
    background: $ssb-green-2;
    bottom: 0;
    height: 0;
    left: 0;
    position: absolute;
    transition: height .2s;
    width: 100%;
    z-index: -1;
  }

  &:hover {
    .animate-background {
      height: 100%;
    }
  }

  .info-box {
    background: $ssb-green-5;
    color: $ssb-white;
    max-width: 250px;
    padding: 18px;
    position: absolute;

    .close-button {
      right: 4px;
      position: absolute;
      top: 4px;
    }

    .info-text {
      color: $ssb-white;
      font-size: 12px;
    }
  }
}