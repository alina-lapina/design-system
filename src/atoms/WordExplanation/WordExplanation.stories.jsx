import React from 'react';
import { storiesOf } from '@storybook/react';
import centered from '@storybook/addon-centered/react';
import WordExplanation from './WordExplanation';

const placeholderText = `This is the explanation of the word
written in one or more paragraphs depending on the amount of space needed`;

storiesOf('Atoms|WordExplanation', module).addDecorator(centered)
	.add('As a wrapper', () => (
		<span>Lorem ipsum dolor sit amet, consectetur adipiscing elit.
			Morbi laoreet rhoncus lorem, eu finibus diam interdum sed.
			Integer eleifend, nunc nec venenatis lacinia, tortor ex luctus est, sed tincidunt nulla turpis eget quam.
			Nullam elit nunc, rutrum quis viverra eget, varius vel neque
			rutum <WordExplanation explanation={placeholderText}>only this is wrapped!</WordExplanation>
			Ut gravida orci bibendum dolor dictum, id pretium quam bibendum.
			Vivamus nunc enim, aliquam et risus in, hendrerit pulvinar magna.
			Aenean tincidunt id est vitae viverra. Maecenas non sem et sapien faucibus accumsan nec ut ipsum.
		</span>
	));
