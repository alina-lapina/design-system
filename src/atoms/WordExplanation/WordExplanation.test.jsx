import React, { useEffect } from 'react';
import { mount, shallow } from 'enzyme';
import WordExplanation from './WordExplanation';

const placeholderText = `This is the explanation of the word
written in one or more paragraphs depending on the amount of space needed`;

const mockUseEffect = () => useEffect.mockImplementationOnce(f => f());

describe('WordExplanation component', () => {
	test('Matches the snapshot', () => {
		const wrapper = shallow(<WordExplanation explanation={placeholderText}>Something else</WordExplanation>);
		expect(wrapper).toMatchSnapshot();
	});
	test('Info box should not exist on init', () => {
		const wrapper = shallow(<WordExplanation explanation={placeholderText}>Text and stuff</WordExplanation>);
		expect(wrapper.exists('.info-box')).toEqual(false);
	});
	test('Info box should render on click', () => {
		const wrapper = shallow(<WordExplanation explanation={placeholderText}>Text and stuff</WordExplanation>);
		wrapper.find('.word-explanation-wrap').simulate('click');
		expect(wrapper.exists('.info-text')).toEqual(true);
	});
	// TODO: Test hooks and
});
