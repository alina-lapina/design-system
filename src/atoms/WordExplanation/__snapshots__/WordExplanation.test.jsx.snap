// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WordExplanation component Matches the snapshot 1`] = `
<span
  className="word-explanation-wrap"
  onClick={[Function]}
>
  Something else
  <i />
  Â 
  <div
    className="animate-background"
  />
</span>
`;
